coord/src/rmr/node_map.c

```diff
@@ -1,110 +1,84 @@
+ * the Server Side Public License v1 (SSPLv1).
+ */
 
 #include <stdlib.h>
+#include <stdio.h>
+
+#include "node_map.h"
+
+#define ADDRESS_LENGTH 256
+
+struct MRNodeMap {
+  dict *nodes;
+  dict *hosts;
+};
 
 void MRNodeMapIterator_Free(MRNodeMapIterator *it) {
-  TrieMapIterator_Free(it->iter);
+  dictReleaseIterator(it->iter);
 }
 
-MRClusterNode *_nmi_allNext(MRNodeMapIterator *it) {
-  char *str;
-  tm_len_t len;
-  void *p;
-  if (!TrieMapIterator_Next(it->iter, &str, &len, &p)) {
-    return NULL;
-  }
-  return p;
+static MRClusterNode *_nmi_allNext(MRNodeMapIterator *it) {
+  dictEntry *de = dictNext(it->iter);
+  return de ? dictGetVal(de) : NULL;
 }
 
-MRClusterNode *_nmi_randomNext(MRNodeMapIterator *it) {
-  char *host;
-  tm_len_t len;
-  void *p;
-  if (!TrieMapIterator_Next(it->iter, &host, &len, &p)) {
-    return NULL;
-  }
-  int retries = 0;
-  MRClusterNode *n = NULL;
-  while (retries < 3) {
-    n = TrieMap_RandomValueByPrefix(it->m->nodes, host, len);
-    if (!n) break;
-    // do not select the same node as excluded
-    if (n->flags & MRNode_Self) {
-      retries++;
-      continue;
-    }
-    break;
-  }
-  return n;
-}
 MRNodeMapIterator MRNodeMap_IterateAll(MRNodeMap *m) {
-  return (MRNodeMapIterator){
-      .Next = _nmi_allNext, .m = m, .excluded = NULL, .iter = TrieMap_Iterate(m->nodes, "", 0)};
-}
-
-MRNodeMapIterator MRNodeMap_IterateHost(MRNodeMap *m, const char *host) {
   return (MRNodeMapIterator){.Next = _nmi_allNext,
                              .m = m,
                              .excluded = NULL,
-                             .iter = TrieMap_Iterate(m->nodes, host, strlen(host))};
+                             .iter = dictGetIterator(m->nodes),
+                             .host = NULL};
 }
-MRNodeMapIterator MRNodeMap_IterateRandomNodePerhost(MRNodeMap *m, MRClusterNode *excludeNode) {
-  return (MRNodeMapIterator){.Next = _nmi_randomNext,
-                             .m = m,
-                             .excluded = excludeNode,
-                             .iter = TrieMap_Iterate(m->hosts, "", 0)};
+
+/* Return 1 if this is the host of the node's endpoint */
+static int IsNodeHost(const MRClusterNode *node, const char *host) {
+  return !strcasecmp(node->endpoint.host, host);
 }
 
-void *_node_replace(void *oldval, void *newval) {
-  return newval;
+static MRClusterNode *_nmi_hostNext(MRNodeMapIterator *it) {
+  dictEntry *de;
+  while ((de = dictNext(it->iter)) && !IsNodeHost(dictGetVal(de), it->host)) {
+  }
+  return de ? dictGetVal(de) : NULL;
 }
 
-void _nodemap_free(void *ptr) {
-  // do not delete anything - the object is allocated elsewhere
+MRNodeMapIterator MRNodeMap_IterateHost(MRNodeMap *m, const char *host) {
+  return (MRNodeMapIterator){.Next = _nmi_hostNext,
+                             .m = m,
+                             .excluded = NULL,
+                             .iter = dictGetIterator(m->nodes),
+                             .host = host};
```

Comment:
Do we need to copy the host here? IIUC we did not hold it before but just used it to find the subtrie to iterate on,  is that right?

Reply:
right, so if we want to iterate a specific host with a dictionary we need to check each entry returned by the addresses dictionary iterator